### DISCUSSION TITLE
Need to sign binaries inside gauge_bin folder in postbuild action.

### DISCUSSION THREAD

**Discussion by ayankumaar** [getgauge/gauge#1956] [General]
I have the following gauge setup in my system.
Gauge version: 1.1.7

Plugins
-------
dotnet (0.3.1)
html-report (4.0.12)
screenshot (0.0.1)

While I am trying to run my script the binaries inside the gauge_bin folder are getting blocked. I need to code sign those to unblock. How can I do that in post build action with gauge ?

**Comment by sriv** [comment]
Gauge does not have a bost build hook (yet). But you can have your own build system and point to the binaries by setting `GAUGE_CUSTOM_BUILD_PATH`

**Reply by ayankumaar** [reply]
@sriv  Should we update GAUGE_CUSTOM_BUILD_PATH  under the "env" folder in the default.properties file at the project level ? or we should set it as an environment variable ?

**Reply by sriv** [reply]
Env variable works (I've tried it). Properties - I _think_ it should work, but I haven't tried it.

**Reply by ayankumaar** [reply]
@sriv Even after setting GAUGE_CUSTOM_BUILD_PATH as an environment variable while executing the tests gauge is creating the gauge_bin folder at the project level. I want to prevent the gauge_bin folder from getting created. How to achieve that ?

**Reply by ayankumaar** [reply]
@sriv  is there any update on this ?

**Reply by sriv** [reply]
Sorry, got distracted into other things.

I just tried with:

```
Gauge version: 1.2.1
Commit Hash: fa7918b5

Plugins
-------
dotnet (0.3.1)
html-report (4.0.12)
screenshot (0.0.1)
```

And I can see that setting `GAUGE_CUSTOM_BUILD_PATH` skips building the project and does not create `gauge_bin` for me. I am on windows 10.

**Reply by sriv** [reply]
And I tried setting environment variable in powershell, as well as setting `GAUGE_CUSTOM_BUILD_PATH` as a property in `env\default\default.properties`. Both options worked for me.

**Reply by ayankumaar** [reply]
@sriv I have setup GAUGE_CUSTOM_BUILD_PATH at env\default\default.properties. I have tried with both absolute and relative path to the binaries. For ex. I have setup the path like: C:\\MyProj\\Workspace\\MyProj\\bin\\Debug\\netcoreapp3.1 and as well as bin\\Debug\\netcoreapp3.1.

But, after setting the path I keep getting the error **the gauge server crashed 5 times**

I am on the following version:

Gauge version: 1.2.1
Commit Hash: fa7918b

Plugins
-------
dotnet (0.4.1)
html-report (4.0.12)
screenshot (0.0.1)

**Comment by sriv** [comment]
Do you see any errors in the logs? `the gauge server crashed 5 times` is thrown by vscode, and it indicates that gauge could not run on your project.

Try running `gauge daemon --lsp`  in your terminal and see if you are seeing any errors there.

**Reply by ayankumaar** [reply]
@sriv do I need to copy the spec files, properties, env and manifest files to the GAUGE_CUSTOM_BUILD_PATH ?

**Reply by ayankumaar** [reply]
@sriv is there any update on the latest query ?

**Reply by sriv** [reply]
> do I need to copy the spec files, properties, env and manifest files to the GAUGE_CUSTOM_BUILD_PATH ?

No, only the C# binaries need to be present in the location specified by GAUGE_CUSTOM_BUILD_PATH.

**Reply by ayankumaar** [reply]
@sriv can I set the GUAGE_CUSTOM_BUILD_PATH as a process level environment variable ? The same process which executes my gauge scripts, can I set it in that process itself ? Will it work that way ?

I already tried to set it in the same process. But, I am seeing the gauge_bin folder being generated.

**Reply by sriv** [reply]
`GUAGE_CUSTOM_BUILD_PATH` should be `GAUGE_CUSTOM_BUILD_PATH` - can you try with this?

To answer the question, yes this property can be set as an env var. Note that it has to be set before gauge is launched.

**Reply by ayankumaar** [reply]
@sriv Sorry for the typo. But I have set it correctly in my code.

When we set it as an environment variable, does it have to be as a user level environment variable ?

The other way my questions is, there are three types of environment varaibles:
  1)Process Level
   2) User Level
   3) Machine level

Does it have to be any one among these types ? or it can be any of these ?

**Reply by sriv** [reply]
It looks for process environment variable.

**Reply by ayankumaar** [reply]
@sriv when I set the GAUGE_CUSTOM_BUILD_PATH in the default.properties file, I am seeing the the gauge server crashed 5 times. When I run gauge daemon --lsp I am not seeing anything.  I checked the lsp.log file. I found the following logs:


25-11-2021 11:27:06.477 [Gauge] [INFO] Watching directory: x:\TestScripts\GaugeTestScript\workspace\GaugeTestScript
25-11-2021 11:27:06.483 [Gauge] [DEBUG] Found 2 files
25-11-2021 11:27:06.861 [Gauge] [INFO] Initializing concepts cache with 0 concepts
25-11-2021 11:27:06.866 [Gauge] [INFO] Initializing specs cache with 1 specs
25-11-2021 11:27:06.866 [Gauge] [DEBUG] Adding specs from x:\TestScripts\GaugeTestScript\workspace\GaugeTestScript\Specs\Samplespec.spec
25-11-2021 11:27:06.866 [Gauge] [INFO] Initializing steps cache with 1 steps
25-11-2021 11:27:06.885 [Gauge] [INFO] Starting language runner
25-11-2021 11:27:09.092 [Gauge] [INFO] LangServer: reading on stdin, writing on stdout
25-11-2021 11:27:09.097 [Gauge] [CRITICAL] notif #0: initialized: runtime error: invalid memory address or nil pointer dereference
goroutine 52 [running]:
runtime/debug.Stack(0xc0003c1780, 0xa19d40, 0xfd78e0)
	X:/xxxxtoolca***/win****/go/1.16.4/x64/src/runtime/debug/stack.go:24 +0xa5
github.com/getgauge/gauge/api/lang.recoverPanic(0xc0003960f0)
	X:/a/gauge/gauge/api/lang/server.go:308 +0x65
panic(0xa19d40, 0xfd78e0)
	X:/xxxxtoolca***/win****/go/1.16.4/x64/src/runtime/panic.go:965 +0x1c7
github.com/getgauge/gauge/runner.(*GrpcRunner).ExecuteMessageWithTimeout(...)
	X:/a/gauge/gauge/runner/grpcRunner.go:195
github.com/getgauge/gauge/api/lang.globPatternRequest(0x4, 0x4, 0x9be880)
	X:/a/gauge/gauge/api/lang/runner.go:87 +0x7e
github.com/getgauge/gauge/api/lang.registerRunnerCapabilities(0xbe8ba0, 0xc0003b0000, 0xbeb690, 0xc00009e088, 0x0, 0x0)
	X:/a/gauge/gauge/api/lang/capabilities.go:110 +0x54
github.com/getgauge/gauge/api/lang.(*LangHandler).Handle(0x1035340, 0xbeb690, 0xc00009e088, 0xbe8ba0, 0xc0003b0000, 0xc0003960f0, 0xc000481f08, 0x2f6066, 0xc000087980, 0xc000115380)
	X:/a/gauge/gauge/api/lang/server.go:76 +0x649
github.com/getgauge/gauge/api/lang.(*LangHandler).handle(0x1035340, 0xbeb690, 0xc00009e088, 0xc0003b0000, 0xc0003960f0, 0x0, 0x0, 0x0, 0x0)
	X:/a/gauge/gauge/api/lang/server.go:60 +0xac
github.com/sourcegraph/jsonrpc2.(*HandlerWithErrorConfigurer).Handle(0xc0003a0050, 0xbeb690, 0xc00009e088, 0xc0003b0000, 0xc0003960f0)
	X:/XXXX/runneradmin/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.0.0-20200429184054-15c2290dcb37/handler_with_error.go:21 +0x79
created by github.com/getgauge/gauge/api/lang.lspHandler.Handle
	X:/a/gauge/gauge/api/lang/server.go:55 +0x72

**Reply by sriv** [reply]
Does this happen on a specific project? I am not able to see this issue in a fresh gauge-dotnet project in my machine. I can investigate if there's a way for me to replicate the issue.

Also try running `gauge daemon --lsp --log-level=debug` to see if there's more information

**Reply by ayankumaar** [reply]
@sriv I created a fresh dotnet project in my system. There also, I am noticing the same error.

If I put the GAUGE_CUSTOM_BUILD_PATH in double quotes then I see **"Error: runtime error: invalid memory address or nil pointer dereference"**.

If I don't put it in double quotes then I am seeing **"failed to initialize suite datastore. Error: Exception was thrown by handler."**

I am setting the GAUGE_CUSTOM_BUILD_PATH in the default.properties file under the env folder.

I tried running gauge daemon --lsp --log-level=debug. That doesn't show anything in the console.

**FYI, when I don't set the GAUGE_CUSTOM_BUILD_PATH then it runs fine.**

I am using the following gauge version:

Gauge version: 1.2.1
Commit Hash: fa7918b

Plugins
-------
dotnet (0.4.1)
html-report (4.0.12)
screenshot (0.0.1)

Gauge Extension Version: 0.0.20
VS Code Version: 1.57.0
OS: Windows 10
.NET CORE Version: .NET CORE 3.0 and .NET CORE 3.1 (Have tried with both)

**Reply by sriv** [reply]
I am using:

```
>  dotnet --version
5.0.202
> gauge -v
Gauge version: 1.4.1
Commit Hash: a82478a

Plugins
-------
dotnet (0.4.1)
html-report (4.1.2)
```

See attached sample project, it has `GAUGE_CUSTOM_BUILD_PATH` set to `custom_gauge_bin`
[bar.zip](https://github.com/getgauge/gauge/files/7611821/bar.zip)

I noticed an issue, but that is unrelated to your problem - in vscode, steps are marked as having "duplicate implementation" because it picks one implementation from the source code, and also scans the custom build path for implementations.

**Reply by ayankumaar** [reply]
@sriv when are you upgrading to .NET 6.0 ?

**Reply by ayankumaar** [reply]
@sriv I have been able to set the custom_gauge_bin path in the dotnet.properties file and I am able to trigger an execution from that path. Also, it is not creating the gauge_bin any more. However, I am facing few new issues with this:

1) When I click on **debug scenario** in VS Code then I am **unable to debug** if the custom_gauge_bin path is set.

2) When I click on a step definition and then click "Go to definition", I am getting the following error **"implementation source not found: Step implementation referred from an external project or library**".

3) Also, I am seeing red marks in each step which says **"Duplicate step"**.

I have tried with gauge 1.4.2, 1.4.1 and 1.2.1 with all three I am getting the same issues. Can you please tell me about the possible resolution of these issues.

I am running the following:

dotnet-plugin: 0.4.1
html-plugin: 4.0.12
gauge extension: 0.0.20/0.0.23
dotnet sdk: 5.0.202
VS Code: 1.63.0

**Reply by sriv** [reply]
Yes, all three issues are known consequences of using pre built binaries. Unfortunately there isn't a way for this (and also needs some investigation for making this possible). I would recommend not using custom build path when authoring, but use it in CI etc environments, if that is feasible.

**Reply by ayankumaar** [reply]
@sriv I have removed the gauge_custom_build_path. Most of the things are working as expected after that. 
However, sometimes I am seeing **red marks in each step and it says "rpc error: code = Unknown desc = Exception was thrown by handler"**. What may be causing that issue ? How do I fix that ?

**Reply by ayankumaar** [reply]
@sriv is there any update on my query ?

**Reply by sriv** [reply]
> "rpc error: code = Unknown desc = Exception was thrown by handler"

Usually caused because the runner crashed. Hard to determine why without a way to replicate this. Check logs to see if there's any error.

**Reply by ayankumaar** [reply]
> @sriv I have been able to set the custom_gauge_bin path in the dotnet.properties file and I am able to trigger an execution from that path. Also, it is not creating the gauge_bin any more. However, I am facing few new issues with this:
> 
> 1. When I click on **debug scenario** in VS Code then I am **unable to debug** if the custom_gauge_bin path is set.
> 2. When I click on a step definition and then click "Go to definition", I am getting the following error **"implementation source not found: Step implementation referred from an external project or library**".
> 3. Also, I am seeing red marks in each step which says **"Duplicate step"**.
> 
> I have tried with gauge 1.4.2, 1.4.1 and 1.2.1 with all three I am getting the same issues. Can you please tell me about the possible resolution of these issues.
> 
> I am running the following:
> 
> dotnet-plugin: 0.4.1 html-plugin: 4.0.12 gauge extension: 0.0.20/0.0.23 dotnet sdk: 5.0.202 VS Code: 1.63.0

@sriv are these issues fixed if we use GAUGE_CUSTOM_BUILD_PATH ?

**Reply by ayankumaar** [reply]
@sriv is there any update on my query ?

**Reply by sriv** [reply]
@ayankumaar we seem to be going around in circles. I had [already mentioned](https://github.com/getgauge/gauge/discussions/1956#discussioncomment-1866876) that these are side effects of using GAUGE_CUSTOM_BUILD_PATH and there is no workaround at the moment.

**Reply by ayankumaar** [reply]
@sriv I got what you said. I was expecting a fix to this because these are existing issues. That is why I was checking if there is a fix available.

**Reply by sriv** [reply]
@ayankumaar - no, there is no fix for this yet.

**Reply by ayankumaar** [reply]
@sriv Got it thanks.
